### ISSUE TITLE
New Startup Procedure Concept

### ISSUE THREAD

**Issue by theonemcdonald** [theonemcdonald/pfSense-pkg-WireGuard#27]
I have been retooling the interface building and teardown process, because frankly I just wasn't happy with it. Though the idea of building temporary loopback interfaces was clever, it was really a hack. 

I was really trying to keep wireguard interfaces named using the traditional wgX prefix. However, doing so requires these special tricks to get the interface built early during startup. 

I was reminded of an alternative solution from an old project attempting to make a WireGuard package around `wireguard-go` that never really took off (you can find it here: https://github.com/Ascrod/pfSense-pkg-wireguard/) In this project, the author names WireGuard interfaces using the `tun` prefix. pfSense is already plumbed to ignore tun interfaces in the interface mismatch routine during system startup!

The relevant code is here from (https://github.com/pfsense/pfsense/blob/6bb8cdd4d8b892bcb77163c02902d83c26cbe2f2/src/etc/inc/util.inc)

You can see a list of many interface prefixes that are ignored and cannot trigger an interface mismatch during system startup. Because of this, all WireGuard tunnels will be named using the prefix `tun_wg` instead of `wg`. This eliminates the need for any earlyshellcmds as we can now build the tunnels when packages are started towards the end of the bootup process!

```
function is_interface_mismatch() {
	global $config, $g;

	$do_assign = false;
	$i = 0;
	$missing_interfaces = array();
	if (is_array($config['interfaces'])) {
		foreach ($config['interfaces'] as $ifname => $ifcfg) {
			if (interface_is_vlan($ifcfg['if']) != NULL ||
			    interface_is_qinq($ifcfg['if']) != NULL ||
			    preg_match("/^enc|^cua|^tun|^tap|^l2tp|^pptp|^ppp|^ovpn|^ipsec|^gif|^gre|^lagg|^bridge|vlan|_wlan|_\d{0,4}_\d{0,4}$/i", $ifcfg['if'])) {
				// Do not check these interfaces.
				$i++;
				continue;
			} else if (does_interface_exist($ifcfg['if']) == false) {
				$missing_interfaces[] = $ifcfg['if'];
				$do_assign = true;
			} else {
				$i++;
			}
		}
	}

	if (file_exists("{$g['tmp_path']}/assign_complete")) {
		$do_assign = false;
	}

	if (!empty($missing_interfaces) && $do_assign) {
		file_put_contents("{$g['tmp_path']}/missing_interfaces", implode(' ', $missing_interfaces));
	} else {
		@unlink("{$g['tmp_path']}/missing_interfaces");
	}

	return $do_assign;
}
```

**Comment by vajonam** [comment]
will wait for this change before i try to install. how can I remove the `earlyshellcmd` and `rc.bootstrap_wireguard` or will the next release include some code to clean this up?

**Comment by theonemcdonald** [comment]
Most likely won't cleanup. Because this is still experimental, I don't see the value of moving old configuration forward until we are in the Netgate repository. For now, just edit your config.xml manually by removing the `<earlyshellcmd>` entries and deleting `rc.bootstrap_wireguard` from `/usr/local/etc/`. That should clean it up.

**Comment by FineWolf** [comment]
> Most likely won't cleanup. 

In that case, make sure this is clearly indicated in the release notes; no one will search old issues or changes in the README.md to detect those kind of changes.

**Comment by theonemcdonald** [comment]
It is something I'm going to address in my next progress report video, which I'm about to film now.

I hate pulling breaking schema changes like this, but I'm confident that this will be the last breaking change moving forward, unless someone can demonstrate a better way of building interfaces at startup. I really like the `tun` prefix because it doesn't rely on anything else than what pfSense provides is, is inherently portable between hardware platforms, and is way more reliable. I think that is worth a bit of some inconvenience here to move to the new paradigm.

**Comment by vajonam** [comment]
hmm.. i have only ever edited the config.xml after an backup and restored it back.. didn't know it can be edited inline. I guess I can see how you insert stuff into it.

**Comment by theonemcdonald** [comment]
Diagnostics > Edit File > /conf/config.xml

You can make changes and save. Now it's possible that the very next interaction you have with pfSense might read from the `$config` array and then write that to disk, thus overwriting your edits...I've seen that a few times, but it should be very rare. There are enough calls to `parse_config` scattered throughout pfSense internals that the global `$config` array should be up-to-date most likely.

**Comment by SPCulhane** [comment]
After manually editing your /config/config.xml. Delete the cache file, /tmp/config.cache and refresh your browser. This will create a new cache file based on your edits.

**Comment by theonemcdonald** [comment]
@SPCulhane Good call, thanks for the tip!

**Comment by Tigger2014** [comment]
I like this idea also, much cleaner then the various ways we thought about getting around this.

However, do you think we could have some code in the next release post install to migrate from wg to tun_wg for existing users. Yes it's still a breaking change but also a seemless one.


Additionally this code is worth having in the future if we need to change back to wg if netgate add that to the list of exceptions.

Failing that I also agree it would be really useful to have the steps on the next release to adjust our installs for this change.

**Comment by SPCulhane** [comment]
> 
> 
> @SPCulhane Good call, thanks for the tip!

@theonemcdonald Anytime! I use that trick a lot. Thank you for putting this package together!! After upgrading to 2.5.1 and copying over my old keys, it worked right out of the box! I still have to create an Interface, GW, NAT rule so internet traffic can route back to an endpoint. I had to do that when I used wireguard-go months ago before 2.5. I would prefer not to since I'm only using it to access my network externally but I understand the limitations.
